File: kotlin-server/server/src/main/kotlin/ccr/server/net/SpecialEffectsGameObj.kt
package ccr.server.net

import ccr.server.defs.SpecialEffectsGameObjDef

// C++: SpecialEffectsGameObj : public PhysicalGameObj (specialeffectsgameobj.cpp)
// Hierarchy: NetworkObject → BaseGameObj → ScriptableGameObj → DamageableGameObj
//            → PhysicalGameObj → SpecialEffectsGameObj
open class SpecialEffectsGameObj() : PhysicalGameObj() {

    // C++: float LifeRemaining (initialized to -1.0F)
    var lifeRemaining: Float = -1.0f

    // C++: bool IsInitialized (initialized to false)
    var isInitialized: Boolean = false

    init {
        // C++: SpecialEffectsGameObj() — no additional init beyond field defaults
    }

    // Secondary constructor for tests — bypasses Init() / definition pipeline.
    constructor(
        definitionId: Int,
        position: ccr.math.Vector3 = ccr.math.Vector3(),
        modelName: String = "",
    ) : this() {
        definition = ccr.server.defs.BaseGameObjDef(
            name = "sfx_$definitionId", id = definitionId.toUInt(), chunkId = 0u
        )
        this.position = position
        this.modelName = modelName
    }

    // C++: ~SpecialEffectsGameObj() — no extra cleanup
    override fun destruct() {
        super.destruct()
    }

    // C++: virtual void Init()
    override fun init() {
        init(getSpecialEffectsDefinition())
    }

    // C++: void Init(const SpecialEffectsGameObjDef & definition)
    fun init(definition: SpecialEffectsGameObjDef) {
        super.init(definition)

        // C++: Make sure collisions are turned off on this effect
        val physObj = peekPhysicalObject()
        if (physObj != null) {
            physObj.setCollisionGroup(UNCOLLIDEABLE_GROUP)
        }
    }

    // C++: const SpecialEffectsGameObjDef & Get_Definition() const
    fun getSpecialEffectsDefinition(): SpecialEffectsGameObjDef =
        definition as SpecialEffectsGameObjDef

    // C++: void Do_Effect()
    protected fun doEffect() {
        // C++: Make sure we have a render object
        val model = peekModel() ?: return

        // C++: Should we play an animation?
        val animName = getSpecialEffectsDefinition().animationName
        if (animName.isNotEmpty()) {
            // C++: HAnimClass* animation = WW3DAssetManager::Get_Instance()->Get_HAnim(anim_name)
            //      if (animation != NULL) { LifeRemaining = animation->Get_Total_Time(); ... }
            val animation = WW3DAssetManager.getInstance().getHAnim(animName)
            if (animation != null) {
                lifeRemaining = animation.getTotalTime()
                (model as? Animatable3DObjClass)?.setAnimation(animation, 0f, RenderObjClass.ANIM_MODE_ONCE)
                // REF_PTR_RELEASE(animation) — GC handles in Kotlin
            }
        }

        // C++: Should we play a sound?
        if (getSpecialEffectsDefinition().soundDefId > 0) {
            WWAudioClass.getInstance().createInstantSound(
                getSpecialEffectsDefinition().soundDefId,
                getTransform(),
            )
        }
    }

    // C++: void Think()
    override fun think() {
        // C++: Kick off the special effect
        if (!isInitialized) {
            doEffect()
            isInitialized = true
        }

        // C++: Check to see if we should make ourselves go away
        if (lifeRemaining >= 0f) {
            lifeRemaining -= TimeManager.getFrameSeconds()
            if (lifeRemaining <= 0f) {
                setDeletePending()
            }
        }
    }

    // C++: bool Save(ChunkSaveClass & csave)
    override fun save(csave: ChunkSaveClass): Boolean {
        csave.beginChunk(CHUNKID_PARENT)
        super.save(csave)
        csave.endChunk()

        csave.beginChunk(CHUNKID_VARIABLES)
        saveVariables(csave)
        csave.endChunk()

        return true
    }

    // C++: void Save_Variables(ChunkSaveClass & csave)
    protected fun saveVariables(csave: ChunkSaveClass) {
        csave.writeMicroChunk(VARID_LIFE_REMAINING, lifeRemaining)
        csave.writeMicroChunk(VARID_IS_INITIALIZED, isInitialized)
    }

    // C++: bool Load(ChunkLoadClass & cload)
    override fun load(cload: ChunkLoadClass): Boolean {
        while (cload.openChunk()) {
            when (cload.curChunkId) {
                CHUNKID_PARENT    -> super.load(cload)
                CHUNKID_VARIABLES -> loadVariables(cload)
                else -> error("Unrecognized SpecialEffectsGameObj chunk ID: ${cload.curChunkId}")
            }
            cload.closeChunk()
        }
        return true
    }

    // C++: void Load_Variables(ChunkLoadClass & cload)
    protected fun loadVariables(cload: ChunkLoadClass) {
        while (cload.openMicroChunk()) {
            when (cload.curMicroChunkId) {
                VARID_LIFE_REMAINING -> lifeRemaining = cload.readFloat()
                VARID_IS_INITIALIZED -> isInitialized = cload.readBool()
                else -> error("Unrecognized SpecialEffectsGameObj variable chunk ID: ${cload.curMicroChunkId}")
            }
            cload.closeMicroChunk()
        }
    }

    companion object {
        // C++: CHUNKID_GAME_OBJECT_SPECIAL_EFFECTS = 0x0004012F (combatchunkid.h)
        const val NETWORK_CLASS_ID: Int = 0x0004012F  // CHUNKID_GAME_OBJECT_SPECIAL_EFFECTS

        // C++: specialeffectsgameobj.cpp local enum starting at CHUNKID_PARENT = 0x09010236
        private const val CHUNKID_PARENT    = 0x09010236
        private const val CHUNKID_VARIABLES = 0x09010237

        // C++: micro-chunk IDs for Save_Variables / Load_Variables
        private const val VARID_LIFE_REMAINING = 1
        private const val VARID_IS_INITIALIZED = 2

        // C++: UNCOLLIDEABLE_GROUP = 1 (combat.h typedef enum, combat.h:42) — collides with nothing
        private const val UNCOLLIDEABLE_GROUP = 1
    }
}

File: original/Code/Combat/specialeffectsgameobj.cpp
#include "global.h"
#include "specialeffectsgameobj.h"
#include "combatchunkid.h"
#include "wwaudio.h"
#include "hanim.h"
#include "assetmgr.h"
#include "persistfactory.h"
#include "simpledefinitionfactory.h"
#include "chunkio.h"
#include "rendobj.h"
#include "timemgr.h"
#include "phys.h"
#include "combat.h"

///////////////////////////////////////////////////////////////////////////
//	Persist and Editable framework support
///////////////////////////////////////////////////////////////////////////
SimplePersistFactoryClass	<SpecialEffectsGameObj,		CHUNKID_GAME_OBJECT_SPECIAL_EFFECTS>		_SpecialEffectsGameObjPersistFactory;
SimplePersistFactoryClass	<SpecialEffectsGameObjDef, CHUNKID_GAME_OBJECT_DEF_SPECIAL_EFFECTS>	_SpecialEffectsGameObjDefPersistFactory;
DECLARE_DEFINITION_FACTORY	(SpecialEffectsGameObjDef, CLASSID_GAME_OBJECT_DEF_SPECIAL_EFFECTS, "Special Effects") _SpecialEffectsGameObjDefDefFactory;

///////////////////////////////////////////////////////////////////////////
//	Constants
///////////////////////////////////////////////////////////////////////////
enum
{
	CHUNKID_DEF_PARENT				=	0x09010212,
	CHUNKID_DEF_VARIABLES,

	VARID_DEF_ANIMATION_NAME		= 1,
	VARID_DEF_SOUNDID,
};

enum
{
	CHUNKID_PARENT						=	0x09010236,
	CHUNKID_VARIABLES,

	VARID_LIFE_REMAINING				= 1,
	VARID_IS_INITIALIZED
};

///////////////////////////////////////////////////////////////////////////
//
//	SpecialEffectsGameObjDef
//
///////////////////////////////////////////////////////////////////////////
SpecialEffectsGameObjDef::SpecialEffectsGameObjDef (void)	:
	SoundDefID (0)
{
	MODEL_DEF_PARAM(SpecialEffectsGameObjDef, PhysDefID, "TimedDecorationPhysDef");
	EDITABLE_PARAM (SpecialEffectsGameObjDef, ParameterClass::TYPE_STRING,					AnimationName);
	EDITABLE_PARAM (SpecialEffectsGameObjDef, ParameterClass::TYPE_SOUNDDEFINITIONID,	SoundDefID);
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	~SpecialEffectsGameObjDef
//
///////////////////////////////////////////////////////////////////////////
SpecialEffectsGameObjDef::~SpecialEffectsGameObjDef (void)
{
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Get_Class_ID
//
///////////////////////////////////////////////////////////////////////////
uint32_t
SpecialEffectsGameObjDef::Get_Class_ID (void) const	
{ 
	return CLASSID_GAME_OBJECT_DEF_SPECIAL_EFFECTS; 
}

///////////////////////////////////////////////////////////////////////////
//
//	Create
//
///////////////////////////////////////////////////////////////////////////
PersistClass *
SpecialEffectsGameObjDef::Create (void) const 
{
	SpecialEffectsGameObj *obj = new SpecialEffectsGameObj;
	obj->Init (*this);
	return obj;
}

///////////////////////////////////////////////////////////////////////////
//
//	Save
//
///////////////////////////////////////////////////////////////////////////
bool
SpecialEffectsGameObjDef::Save (ChunkSaveClass &csave)
{
	csave.Begin_Chunk (CHUNKID_DEF_PARENT);
		PhysicalGameObjDef::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_DEF_VARIABLES);
		Save_Variables (csave);
	csave.End_Chunk ();
	return true;
}

///////////////////////////////////////////////////////////////////////////
//
//	Load
//
///////////////////////////////////////////////////////////////////////////
bool
SpecialEffectsGameObjDef::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ()) {
		switch(cload.Cur_Chunk_ID ()) {

			case CHUNKID_DEF_PARENT:
				PhysicalGameObjDef::Load (cload);
				break;

			case CHUNKID_DEF_VARIABLES:
				Load_Variables (cload);
				break;
	  
			default:
				Debug_Say (("Unrecognized SpecialEffectsGameObjDef chunk ID\n"));
				break;

		}

		cload.Close_Chunk ();
	}

	return true;
}

///////////////////////////////////////////////////////////////////////////
//
//	Save_Variables
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObjDef::Save_Variables (ChunkSaveClass &csave)
{
	WRITE_MICRO_CHUNK_WWSTRING	(csave, VARID_DEF_ANIMATION_NAME,	AnimationName);
	WRITE_MICRO_CHUNK				(csave, VARID_DEF_SOUNDID,				SoundDefID);
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObjDef::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk()) {
		switch (cload.Cur_Micro_Chunk_ID()) {

			READ_MICRO_CHUNK_WWSTRING	(cload, VARID_DEF_ANIMATION_NAME,	AnimationName);
			READ_MICRO_CHUNK				(cload, VARID_DEF_SOUNDID,				SoundDefID);
			
			default:
				Debug_Say (("Unrecognized SpecialEffectsGameObjDef Variable chunkID %d\n", cload.Cur_Micro_Chunk_ID ()));
				break;
		}
		cload.Close_Micro_Chunk ();
	}

	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
///////////////////////////////////////////////////////////////////////////
const PersistFactoryClass &
SpecialEffectsGameObjDef::Get_Factory (void) const 
{ 
	return _SpecialEffectsGameObjDefPersistFactory; 
}

/*
**
**	Start of SpecialEffectsGameObj
**
*/

///////////////////////////////////////////////////////////////////////////
//
//	SpecialEffectsGameObj
//
///////////////////////////////////////////////////////////////////////////
SpecialEffectsGameObj::SpecialEffectsGameObj (void)	:
	LifeRemaining (-1.0F),
	IsInitialized (false)
{
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	~SpecialEffectsGameObj
//
///////////////////////////////////////////////////////////////////////////
SpecialEffectsGameObj::~SpecialEffectsGameObj (void)
{
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Get_Factory
//
///////////////////////////////////////////////////////////////////////////
const PersistFactoryClass &
SpecialEffectsGameObj::Get_Factory (void) const 
{
	return _SpecialEffectsGameObjPersistFactory;
}

////////////////////////////////////////////////////////////////
//
//	Init
//
////////////////////////////////////////////////////////////////
void SpecialEffectsGameObj::Init( void )
{
	Init( Get_Definition() );
}

///////////////////////////////////////////////////////////////////////////
//
//	Init
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObj::Init (const SpecialEffectsGameObjDef &definition)
{
	PhysicalGameObj::Init (definition);

	//
	//	Make sure collisions are turned off on this effect
	//
	PhysClass *phys_obj = Peek_Physical_Object();
	if (phys_obj != NULL) {
		phys_obj->Set_Collision_Group( UNCOLLIDEABLE_GROUP );
	}
	
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Get_Definition
//
///////////////////////////////////////////////////////////////////////////
const SpecialEffectsGameObjDef &
SpecialEffectsGameObj::Get_Definition (void) const
{
	return (const SpecialEffectsGameObjDef &)BaseGameObj::Get_Definition();
}

///////////////////////////////////////////////////////////////////////////
//
//	Save
//
///////////////////////////////////////////////////////////////////////////
bool
SpecialEffectsGameObj::Save (ChunkSaveClass & csave)
{
	csave.Begin_Chunk (CHUNKID_PARENT);
		PhysicalGameObj::Save (csave);
	csave.End_Chunk ();

	csave.Begin_Chunk (CHUNKID_VARIABLES);
		Save_Variables (csave);
	csave.End_Chunk ();

	return true;
}

///////////////////////////////////////////////////////////////////////////
//
//	Load
//
///////////////////////////////////////////////////////////////////////////
bool
SpecialEffectsGameObj::Load (ChunkLoadClass &cload)
{
	while (cload.Open_Chunk ()) {
		switch(cload.Cur_Chunk_ID ()) {

			case CHUNKID_PARENT:
				PhysicalGameObj::Load (cload);
				break;

			case CHUNKID_VARIABLES:
				Load_Variables (cload);
				break;

			default:
				Debug_Say (("Unrecognized SpecialEffectsGameObj chunk ID\n"));
				break;

		}

		cload.Close_Chunk ();
	}

	return true;
}

///////////////////////////////////////////////////////////////////////////
//
//	Do_Effect
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObj::Do_Effect (void)
{
	//
	//	Make sure we have a render object
	//
	RenderObjClass *model = Peek_Model ();

	if (model != NULL) {

		//
		//	Should we play an animation?
		//
		const StringClass &anim_name = Get_Definition ().AnimationName;
		if (anim_name.Get_Length () > 0) {
			
			//
			//	Load the animation
			//
			HAnimClass *animation =  WW3DAssetManager::Get_Instance ()->Get_HAnim (anim_name);
			if (animation != NULL) {
				LifeRemaining = animation->Get_Total_Time ();
				model->Set_Animation (animation, 0, RenderObjClass::ANIM_MODE_ONCE);
				REF_PTR_RELEASE(animation);
			}
		}

		//
		//	Should we play a sound?
		//
		if (Get_Definition ().SoundDefID > 0) {
			WWAudioClass::Get_Instance()->Create_Instant_Sound (Get_Definition ().SoundDefID, Get_Transform ());
		}
	}

	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Save_Variables
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObj::Save_Variables (ChunkSaveClass &csave)
{
	WRITE_MICRO_CHUNK	(csave, VARID_LIFE_REMAINING, LifeRemaining);
	WRITE_MICRO_CHUNK	(csave, VARID_IS_INITIALIZED, IsInitialized);
	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Load_Variables
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObj::Load_Variables (ChunkLoadClass &cload)
{
	while (cload.Open_Micro_Chunk()) {
		switch (cload.Cur_Micro_Chunk_ID()) {

			READ_MICRO_CHUNK (cload, VARID_LIFE_REMAINING, LifeRemaining);
			READ_MICRO_CHUNK (cload, VARID_IS_INITIALIZED, IsInitialized);
			
			default:
				Debug_Say (("Unrecognized SpecialEffectsGameObj Variable chunkID %d\n", cload.Cur_Micro_Chunk_ID ()));
				break;
		}
		cload.Close_Micro_Chunk ();
	}

	return ;
}

///////////////////////////////////////////////////////////////////////////
//
//	Think
//
///////////////////////////////////////////////////////////////////////////
void
SpecialEffectsGameObj::Think (void)
{

	//
	//	Kick off the special effect
	//
	if (IsInitialized == false) {
		Do_Effect ();
		IsInitialized = true;
	}

	//
	//	Check to see if we should make ourselves go away
	//
	if (LifeRemaining >= 0) {
		LifeRemaining -= TimeManager::Get_Frame_Seconds ();
		if (LifeRemaining <= 0) {
			Set_Delete_Pending ();
		}
	}
	
	return ;
}
